File: scope/lib/core/utils/pii_log_sanitizer.dart
import 'dart:convert';
import 'package:crypto/crypto.dart' as crypto;
import 'package:flutter/services.dart';

/// Configurable sanitization modes for PII logging.
enum SanitizationMode {
  /// Masked placeholder format, e.g. `[REDACTED (12 chars)]`
  masked,

  /// Structural metadata format, e.g. `[title_len=12, sha256=a1b2c345]`
  metadata,

  /// Non-sensitive digest format, e.g. `[sha256=a1b2c345]`
  digest,
}

/// Centralized log sanitization helper for redacting raw notification titles,
/// body contents, and sensitive tokens into structural metadata before logging
/// to system Logcat buffers or debug console streams.
class PiiLogSanitizer {
  PiiLogSanitizer._();

  /// Computes a truncated SHA-256 hash digest hex string.
  static String sha256Digest(String? input, {int length = 8}) {
    if (input == null || input.isEmpty) return 'empty';
    try {
      final bytes = utf8.encode(input);
      final digest = crypto.sha256.convert(bytes).toString();
      if (length > 0 && length < digest.length) {
        return digest.substring(0, length);
      }
      return digest;
    } catch (_) {
      return 'error';
    }
  }

  /// Redacts string input into structural metadata based on [mode].
  static String sanitize(
    String? input, {
    SanitizationMode mode = SanitizationMode.masked,
    String? label,
  }) {
    try {
      if (input == null) {
        final lbl = (label != null && label.isNotEmpty) ? '${label}_len' : 'str_len';
        switch (mode) {
          case SanitizationMode.masked:
            return '[REDACTED]';
          case SanitizationMode.digest:
            return '[sha256=null]';
          case SanitizationMode.metadata:
            return '[$lbl=0, sha256=null]';
        }
      }

      final len = input.length;
      switch (mode) {
        case SanitizationMode.masked:
          return '[REDACTED ($len chars)]';
        case SanitizationMode.digest:
          return '[sha256=${sha256Digest(input)}]';
        case SanitizationMode.metadata:
          final lbl = (label != null && label.isNotEmpty) ? '${label}_len' : 'len';
          return '[$lbl=$len, sha256=${sha256Digest(input)}]';
      }
    } catch (_) {
      return '[REDACTED]';
    }
  }

  /// Shortcut to mask string content with character count: `[REDACTED (X chars)]`.
  static String mask(String? input) {
    return sanitize(input, mode: SanitizationMode.masked);
  }

  /// Shortcut to format string into metadata format: `[title_len=X, sha256=Y]`.
  static String toMetadata(String? input, {String? label}) {
    return sanitize(input, mode: SanitizationMode.metadata, label: label);
  }

  /// Sanitizes exception details to prevent leaking PII or raw payload data.
  static String sanitizeException(Object? error) {
    if (error == null) return 'Exception(null)';
    try {
      if (error is PlatformException) {
        final sanitizedMsg = error.message != null
            ? mask(error.message)
            : '[no message]';
        return 'PlatformException(code: ${error.code}, message: $sanitizedMsg)';
      }
      final str = error.toString();
      return mask(str);
    } catch (_) {
      return 'Exception(sanitization_failed)';
    }
  }
}

File: scope/android/app/src/main/kotlin/com/scope/attentions/PiiLogSanitizer.kt
package com.scope.attentions

import java.security.MessageDigest

/**
 * Centralized log sanitization helper for redacting raw PII (notification titles,
 * body content, and sensitive tokens) into structural metadata before logging to
 * system Logcat buffers.
 */
object PiiLogSanitizer {

    enum class SanitizationMode {
        METADATA,
        MASKED,
        DIGEST
    }

    /**
     * Computes a truncated SHA-256 digest hex string for the given text.
     */
    fun sha256Digest(text: String?, length: Int = 8): String {
        if (text.isNullOrEmpty()) return "empty"
        return try {
            val digest = MessageDigest.getInstance("SHA-256").digest(text.toByteArray(Charsets.UTF_8))
            val hexString = StringBuilder()
            for (b in digest) {
                val hex = Integer.toHexString(0xff and b.toInt())
                if (hex.length == 1) hexString.append('0')
                hexString.append(hex)
            }
            if (length > 0 && length < hexString.length) {
                hexString.substring(0, length)
            } else {
                hexString.toString()
            }
        } catch (e: Exception) {
            "error"
        }
    }

    /**
     * Redacts string input into structural metadata according to the specified mode.
     */
    fun sanitize(
        text: String?,
        label: String? = null,
        mode: SanitizationMode = SanitizationMode.METADATA
    ): String {
        return try {
            if (text == null) {
                val lbl = if (!label.isNullOrEmpty()) "${label}_len" else "str_len"
                return when (mode) {
                    SanitizationMode.MASKED -> "[REDACTED]"
                    SanitizationMode.DIGEST -> "[sha256=null]"
                    SanitizationMode.METADATA -> "[$lbl=0, sha256=null]"
                }
            }

            val len = text.length
            when (mode) {
                SanitizationMode.MASKED -> "[REDACTED ($len chars)]"
                SanitizationMode.DIGEST -> "[sha256=${sha256Digest(text)}]"
                SanitizationMode.METADATA -> {
                    val lbl = if (!label.isNullOrEmpty()) "${label}_len" else "len"
                    "[$lbl=$len, sha256=${sha256Digest(text)}]"
                }
            }
        } catch (e: Exception) {
            "[REDACTED]"
        }
    }

    /**
     * Convenience method to mask string content with character count.
     */
    fun mask(text: String?): String {
        return sanitize(text, mode = SanitizationMode.MASKED)
    }

    /**
     * Convenience method to format string as structural metadata.
     */
    fun toMetadata(text: String?, label: String? = null): String {
        return sanitize(text, label = label, mode = SanitizationMode.METADATA)
    }

    /**
     * Convenience method to format captured notification details safely.
     */
    fun sanitizeNotification(packageName: String?, title: String?, content: String? = null): String {
        val pkg = packageName ?: "unknown"
        val titleMetadata = sanitize(title, label = "title")
        return if (content != null) {
            val contentMetadata = sanitize(content, label = "content")
            "$pkg - $titleMetadata - $contentMetadata"
        } else {
            "$pkg - $titleMetadata"
        }
    }

    /**
     * Sanitizes exception details to prevent leaking PII or raw message payloads.
     */
    fun sanitizeException(e: Throwable?): String {
        if (e == null) return "Exception(null)"
        return try {
            val className = e.javaClass.simpleName
            val message = e.message
            if (message.isNullOrEmpty()) {
                className
            } else {
                "$className(message=${mask(message)})"
            }
        } catch (ex: Exception) {
            "Exception(sanitization_failed)"
        }
    }
}
